pkg/controller/operators/adoption_controller.go

```diff
@@ -543,15 +524,13 @@ func (r *AdoptionReconciler) mapToProviders(obj client.Object) (requests []recon
 	return
 }
 
-func flatten(objs []runtime.Object) (flattened []runtime.Object) {
-	for _, obj := range objs {
-		if nested, err := meta.ExtractList(obj); err == nil {
-			flattened = append(flattened, flatten(nested)...)
-			continue
-		}
-
-		flattened = append(flattened, obj)
+// flatten extracts slice of ObjectList to the slice of Object
+func flatten(lists []client.ObjectList) (flattened []client.Object) {
+	for _, list := range lists {
+		_ = meta.EachListItem(list, func(obj runtime.Object) error {
+			flattened = append(flattened, obj.(client.Object))
+			return nil
+		})
```

Comment:
Are we discounting any errors that this inline function may produce? Do we care about handling errors that may occur when iterating over the list of client objects?

Reply:
Hmm, I think it's probably okay.

In this case, the variables are constrained to `[]client.ObjectList`, which we know contains elements that of `[]client.Object`, so we know the assertion would never panic. We also know that the function literal will always return nil.